week-3/server.js

```diff
@@ -0,0 +1,258 @@
+            customersIdArr= result.rows.map(r =>r.id);})
+   
+app.get("/customers", function(req, res) {
+    pool.query('SELECT * FROM customers', (error, result) => {
+        res.json(result.rows);
+    });
+});
+
+app.get("/customers/by_name/:name", function(req, res) {
+    const cudtomerName=req.params.name.toLowerCase();
+    pool.query("SELECT * FROM customers where lower(name) like '%'||$1||'%'",[cudtomerName], (error, result) => {
+        res.json(result.rows);
+    });
+});
+
+app.post("/suppliers" ,function (req,res){
+    const newName=req.body.name;
+    const newCountry=req.body.country;
+
+    pool.query("SELECT 1 FROM suppliers WHERE supplier_name=$1", [newName],
+            (err, result) => {
+      if (result.rowCount > 0) {      // note the use of result.rowCount
+        return res
+          .status(400)
+          .send("A supplier with that name  already exists!");
+      } else {
+    pool.query("insert into suppliers (supplier_name,country )"+
+    "values($1,$2) returning id",[newName ,newCountry],(error,result)=> {
+        if ((error)== undefined){
+            //res.send("New customer added.");
+            //res.json(result.rows[0].id);
+            res.send("New supplier added.");
+        }
+        else {
+        res.status(400).send("wrong request "+error)
+    console.log(error);}
+    });
+}})
+});
+
+// delete order item
+app.delete("/order_items/:id" ,function (req,res){
+    const orderId=parseInt( req.params.id);
+
+    pool.query("delete from order_items  where  id=$1"
+ ,[orderId],(error)=> {
+        if ((error)== undefined){
+            
+            res.send("order_items "+orderId+" deleted.");
+        }
+    });
+});
+app.get("/suppliers", function(req, res) {
+    pool.query('SELECT * FROM suppliers', (error, result) => {
+        res.json(result.rows);
+    });
+});
+// app.get("/products", function(req, res) {
+//     pool.query('select p.product_name,pa.unit_price,s.supplier_name'+
+//     ' from products p join product_availability pa on(p.id=pa.prod_id)'+
+//     'join suppliers s on (pa.supp_id=s.id) ;', (error, result) => {
+//         res.json(result.rows);
+//     });
+// });
+
+
+//Homework
+
+//filter the list of products
+
+app.get("/products", function(req, res) {
+    const searchProduct = req.query.name;
+    console.log(searchProduct);
+    if (searchProduct!=undefined)
+    pool.query("select p.product_name,pa.unit_price,s.supplier_name"+
+    ' from products p join product_availability pa on(p.id=pa.prod_id)'+
+    "join suppliers s on (pa.supp_id=s.id) where lower(p.product_name) like '%'||$1||'%'",[searchProduct.toLowerCase()], (error, result) => {
+        res.json(result.rows);
+    })
+        else
+        pool.query('select p.product_name,pa.unit_price,s.supplier_name'+
+    ' from products p join product_availability pa on(p.id=pa.prod_id)'+
+    'join suppliers s on (pa.supp_id=s.id) ;', (error, result) => {
+        res.json(result.rows);
+    });
```

Comment:
You could avoid repetition by concatenating the "where" clause to the main query, rather than having 2 completely separate strings.